llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -9241,9 +9242,10 @@ LegalizerHelper::lowerMemmove(MachineInstr &MI, Register Dst, Register Src,
     // Don't promote to an alignment that would require dynamic stack
     // realignment.
     const TargetRegisterInfo *TRI = MF.getSubtarget().getRegisterInfo();
-    if (!TRI->hasStackRealignment(MF))
-      while (NewAlign > Alignment && DL.exceedsNaturalStackAlignment(NewAlign))
-        NewAlign = NewAlign.previous();
+    if (NewAlign > Alignment && !TRI->hasStackRealignment(MF)) {
+      const TargetFrameLowering *TFL = MF.getSubtarget().getFrameLowering();
```

Comment:
Can we assert that the TFL value is the same as the datalayout? Preferably we wouldn't have 2, and only have the DL one?

Reply:
I don't like having two of them either, but AIUI the one from DataLayout is more of a heuristic.

Reply:
The TFL one is also another fixed constant, just set in a different place

Reply:
> Can we assert that the TFL value is the same as the datalayout?

It appears they values are different on AMDGPU. Data layout string specifies "S32" (4 bytes), but [TFL value](https://github.com/llvm/llvm-project/blob/fc6300a5f7ef430e4ec86d16be0b146de7fbd16b/llvm/lib/Target/AMDGPU/GCNSubtarget.h#L953) is 16 bytes. I couldn't conclude if this is intended or an oversight.

Reply:
It's not intended. I don't know what the DataLayout one is even used for, it should change to match

Reply:
The DataLayout one is used in similar scenarios, but in middle end passes. I.e. when promoting alloca alignment.